internal/controller/namespacelabel_controller.go

```
@@ -0,0 +1,141 @@
+	"github.com/matanamar10/namespacelabel-operator/internal/finalizer"
+	"github.com/matanamar10/namespacelabel-operator/internal/utils"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// NamespacelabelReconciler reconciles a Namespacelabel object
+type NamespacelabelReconciler struct {
+	client.Client
+	Log      logr.Logger
+	Scheme   *runtime.Scheme
+	Recorder record.EventRecorder
+}
+
+// RBAC permissions required by the reconciler to manage Namespacelabels and namespaces.
+// +kubebuilder:rbac:groups=labels.dana.io,resources=namespacelabels,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=labels.dana.io,resources=namespacelabels/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=labels.dana.io,resources=namespacelabels/finalizers,verbs=update
+// +kubebuilder:rbac:groups="",resources=namespaces,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile reconciles a Namespacelabel object and manages finalizers and namespace labels.
+func (r *NamespacelabelReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// Obtain a logger for this context
+	_ = log.FromContext(ctx)
+
+	var namespaceLabel labelsv1.Namespacelabel
+	if err := r.Get(ctx, req.NamespacedName, &namespaceLabel); err != nil {
+		// Ignore the error if the resource was not found
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if !namespaceLabel.ObjectMeta.DeletionTimestamp.IsZero() {
+		// Clean up resources before deletion
+		if err := finalizer.CleanupFinalizer(ctx, r.Client, &namespaceLabel); err != nil {
+			r.Log.Error(err, "Failed to clean up labels during finalizer")
+			return ctrl.Result{}, err
```

Instead of logging an error, you should return an error with context

https://100go.co/#handling-an-error-twice-52
https://google.github.io/styleguide/go/best-practices#adding-information-to-errors

---

This is true throughout the code, not just here